Get rid of Gitpod layer

csweichel opened the pull request:
fixes #4899 

This PR removes the need for the Gitpod layer, by moving the user and group creation to supervisor.

### How to test
Start workspaces with as many different images you can think of.
Especially interesting: images that don't have a Gitpod user to begin with, e.g. https://github.com/csweichel/test-repo/tree/image-ubuntu

### Open questions:
- do we still need the `.bashrc` modifications? IMHO that was somewhat of an antipattern to begin with and we should find other ways if need be.
- what about images that don't have git. Can we live with that or should we ship a "fallback Git" as part of supervisor?

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4923#" title="Author self-approved">csweichel</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [csweichel]
- ~~[components/supervisor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/OWNERS)~~ [csweichel]
- ~~[components/ws-manager/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

aledbf commented:
> Can we live with that or should we ship a "fallback Git" as part of supervisor?

It would be great if we could use any of the Go libraries available to avoid distribution dependent binaries

csweichel commented:
> > Can we live with that or should we ship a "fallback Git" as part of supervisor?
> 
> It would be great if we could use any of the Go libraries available to avoid distribution dependent binaries

I agree. We have used [go-git](https://github.com/go-git/go-git) in the past and found incompatibilities at the time regarding the global Git config. It's well possible those have been resolved by now.

Now I reckon the question is if this way we could use the latest --filter optimisations :)

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4923#" title="Author self-approved">csweichel</a>*

Associated issue: *#4899*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [csweichel]
- ~~[components/supervisor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/OWNERS)~~ [csweichel]
- ~~[components/ws-manager/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager/OWNERS)~~ [csweichel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

csweichel commented:
/hold

we really need to test this thoroughly with different images

aledbf commented:
/werft run

aledbf commented:
@csweichel I tested all the examples, gitpod, and even your image-ubuntu repository. Everything works.

Just one question, what's the expected behavior in case the gitpod account exists but the UID is different?
example https://github.com/aledbf/gitpod-user-uid/blob/main/.gitpod.Dockerfile

csweichel commented:
> @csweichel I tested all the examples, gitpod, and even your image-ubuntu repository. Everything works.
> 
> Just one question, what's the expected behavior in case the gitpod account exists but the UID is different?
> example https://github.com/aledbf/gitpod-user-uid/blob/main/.gitpod.Dockerfile

Excellent question. Two ways come to mind:
1. the cheap and easy way: fail the workspace startup with a sensible, actionable error message
2. the clever way: produce a user with a different name but correct UID. I don't think we actually rely on Gitpod as name but rather use the UID/GID (might have to double check the content-initializer code).

The first one is clearly much easier to implement. WDYT?

aledbf commented:
> the cheap and easy way: fail the workspace startup with a sensible, actionable error message

:+1: 

(right now stays in `Pulling container image …`)

aledbf commented:
@csweichel awesome.

![Screenshot from 2021-08-02 13-23-42](https://user-images.githubusercontent.com/161571/127900262-88fbf7a5-6b66-4cd7-824f-8323e541b19c.png)

aledbf commented:
/lgtm
/approve

roboquat commented:
LGTM label has been added.  <details>Git tree hash: 3d4baa050cf30fbe095186e76fc63c5cc0ba9365</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4923#issuecomment-891200580" title="Approved">aledbf</a>*

Associated issue: *#4899*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [aledbf]
- ~~[components/supervisor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/OWNERS)~~ [aledbf]
- ~~[components/workspacekit/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/workspacekit/OWNERS)~~ [aledbf]
- ~~[components/ws-manager/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager/OWNERS)~~ [aledbf]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

JohannesLandgraf commented:
This looks like Italy 🇮🇹 🍕<img width="815" alt="CleanShot 2021-08-02 at 19 32 46@2x" src="https://user-images.githubusercontent.com/22498066/127901231-a9f6e173-ee37-40d0-9a35-93d25ddf932d.png"> cc @fntlnz @leodido

jankeromnes commented on components/image-builder-mk3/workspace-image-layer/gitpod-layer/prepend.dockerfile:
```diff
@@ -1,7 +0,0 @@
-
-
-USER ROOT
-COPY ./gitpod-cli /usr/bin/gp
```
> This PR removes the need for the Gitpod layer, by moving the user and group creation to supervisor.

What about the `gp` CLI? Surely we want to continue installing it as well. (Should at least test that it's still there.)

jankeromnes commented on components/image-builder-mk3/workspace-image-layer/gitpod-layer/Dockerfile:
```diff
@@ -1,34 +0,0 @@
-# Switch to user gitpod
-USER gitpod
-
-# Configure user shell
-# TODO Remove this in the near future when we do not need ~/.bashrc appends/prepends any more
-RUN \
-    # REALLY do not print motd
-    touch ~/.hushlogin                                                      && \
-    # Configure shell
-    BASH_RC=~/.bashrc; if [ -f "$BASH_RC" ]; then cp "$BASH_RC" ~/.bashrc-org; else touch ~/.bashrc-org; fi && \
-    cat /var/gitpod/gitpod/.bashrc-prepend > "$BASH_RC"                            && \
-    cat ~/.bashrc-org >> "$BASH_RC"                                         && \
-    cat /var/gitpod/gitpod/.bashrc-append >> "$BASH_RC"
```
Yes! 😍

jankeromnes commented on components/image-builder-mk3/workspace-image-layer/gitpod-layer/debian/gitpod/.bashrc-prepend:
```diff
@@ -1,58 +0,0 @@
-    export GIT_EDITOR="$EDITOR"
-fi
-
-# Workaround Java pre v10 by explicitly setting "-Xmx" for all Hotspot/openJDK VMs
-if [ -n "$GITPOD_MEMORY" ]; then
-    export JAVA_TOOL_OPTIONS="-Xmx${GITPOD_MEMORY}m";
-fi
-
-# Completion for gp command
-. <(gp completion)
-# ide cli config - should be removed when registry facade is default
-if [ ! -d "/ide/bin/" ]; then
-    alias open='gp open'
-    alias code='gp open'
-fi
-
-export GEM_HOME=/workspace/.rvm
-export GEM_PATH=$GEM_HOME:$GEM_PATH
-export PATH=/workspace/.rvm/bin:$PATH
-
-export PIPENV_VENV_IN_PROJECT=true
-export PIP_USER=yes
-export PYTHONUSERBASE=/workspace/.pip-modules
-export PATH=$PYTHONUSERBASE/bin:$PATH
-unset PIP_TARGET
-unset PYTHONPATH
-
-# Set CARGO_HOME to reside in workspace if:
-#  - it's RUNTIME (/workspace present)
-if [ -d /workspace ]; then
-    export CARGO_HOME=/workspace/.cargo
-    export PATH=$CARGO_HOME/bin:$PATH
-fi
-
-export BROWSER="${BROWSER:=gp-preview}"
-
-###############################################################################
-##########  Gitpod - prepend
-###############################################################################
-
```
~~Burn it, burn it all with fire! 🔥🔥🔥~~

TODO:
- [x] `PS1` prompt gets re-defined anyway so line had no effect
- [ ] Test that making a new commit from the console is still okay (`EDITOR` vars)
- [ ] Test that `JAVA_TOOL_OPTIONS` is still defined so that the JVM won't eat up all Gitpod's precious RAM
- [ ] Test `gp` and its auto-completion
- [ ] Test `open` and `code` aliases (nice to haves / but not mandatory)
- [x] `GEM_HOME=/workspace/.rvm` gets overwritten by nearly all Ruby projects anyway
- [x] We want a less crazy-custom Python setup so fine by me
- [ ] Test `cargo install` persistence across restarts
- [x] I've never liked the `BROWSER` automatic behavior (i.e. "automatically open IDE preview when some specific web servers start") (also did that even work in Code?)

I guess most of these changes are fine, because we want fewer non-standard customizations:
- except for `gp` and `JAVA_TOOL_OPTIONS`, which look dangerous and really should be tested
- also, big question: is Full Workspace Backup enabled everywhere now? (If yes, it's fine to remove all `/workspace` hacks -- if not, this PR breaks Cargo, Python and maybe Ruby package persistence)

jankeromnes reviewed (commented):
Whoa, many thanks @csweichel for taking this on! 💯 🎖️ 

<img width="136" alt="Screenshot 2021-08-02 at 22 21 19" src="https://user-images.githubusercontent.com/599268/127919900-4576c806-e1e5-4939-969c-8c844e231bb4.png">

Epic!! 🎸 (Although I'd have preferred fewer lines of new Go code 🙈 looks a bit scary to my untrained eye, as each new line of code has a non-zero risk-of-new-bug -- regardless of how excellent the developer is!)

Added a few comments on the code. Will now test. 😇

jankeromnes commented:
Hmm, I can't seem to open any repository in this deployment 😕

<img width="1440" alt="Screenshot 2021-08-02 at 22 48 13" src="https://user-images.githubusercontent.com/599268/127922299-f08fdc1e-9248-466e-b35e-ab0942fc6159.png">

I tried:

- Clicking `New Workspace` and picking the first template (`template-typescript-node`)
- Opening an [empty repository](https://github.com/jankeromnes/test)
- Opening https://github.com/csweichel/test-repo/tree/image-ubuntu

AlexTugarev commented:
/werft run

AlexTugarev commented:
Notes: 

* new failure mode: `cannot ensure Gitpod user exists: user: unknown user`

roboquat commented:
New changes are detected. LGTM label has been removed.

csweichel commented:
Thanks @jankeromnes  and @AlexTugarev for testing this - and sorry for the obvious blunder.

I've fixed the "unknown user" issue, added `gp` to the path and set `JAVA_TOOL_OPTIONS`.
I have yet to test the `code` and `editor` behaviour, but those should be superseded by registry-facade's behaviour at this point.

AlexTugarev commented:
New workspaces starting just fine. 

The one that failed previously doesn't start. How can that issue be permanent? 

<img width="1006" alt="Screen Shot 2021-08-03 at 12 12 56" src="https://user-images.githubusercontent.com/914497/127998980-ba1373c8-a607-4319-b38f-97409267df5a.png">

csweichel commented:
@AlexTugarev I reckon that's because the "reconciliation loop" of ws-manager-bridge hasn't "stopped" the old workspace yet. It's still stuck in the stopping state from the last attempt

csweichel commented:
/hold cancel

aledbf commented:
/lgtm

roboquat commented:
LGTM label has been added.  <details>Git tree hash: 029093ef5eda27ecc17d9e17c08f7eafd83f4445</details>

roboquat commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4923#issuecomment-892635439" title="LGTM">aledbf</a>*

Associated issue: *#4899*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [aledbf]
- ~~[components/supervisor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/supervisor/OWNERS)~~ [aledbf]
- ~~[components/workspacekit/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/workspacekit/OWNERS)~~ [aledbf]
- ~~[components/ws-manager/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager/OWNERS)~~ [aledbf]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
